nick_work/fdaaa_port.py

```diff
@@ -6,367 +6,310 @@
 from datetime import datetime
 from datetime import timedelta
 from dateutil.relativedelta import relativedelta
-import pandas as pd
 import csv
+from collections import defaultdict
 cs = 'clinical_study'
 
-projectid = "ebmdatalab"
-
-fda_reg_df = pd.read_gbq('''
-SELECT *
-FROM clinicaltrials.jan17_fda_regulation_snapshot 
-''', projectid, dialect = 'standard')
-
-headers = ['nct_id', 'act_flag', 'included_pact_flag', 'has_results', 'pending_results', 'pending_data',\
-           'has_certificate', 'results_due', 'start_date', 'available_completion_date',\
-           'used_primary_completion_date', 'defaulted_pcd_flag', 'defaulted_cd_flag', 'results_submitted_date',\
-           'last_updated_date', 'certificate_date', 'phase', 'enrollment', 'location', 'study_status', 'study_type',\
-           'primary_purpose', 'sponsor', 'sponsor_type', 'collaborators', 'exported', 'fda_reg_drug',\
-           'fda_reg_device', 'is_fda_regulated', 'url', 'title', 'official_title', 'brief_title',\
-           'discrep_date_status', 'late_cert', ' defaulted_date', 'condition', 'condition_mesh', 'intervention',\
-           'intervention_mesh', 'keywords']
+#when actually implemented, point this towards the static location of the file in the cloud
+fda_reg_dict = {}
+with open("fdaaa_regulatory_snapshot.csv") as old_fda_reg:
+    reader = csv.DictReader(old_fda_reg)
+    for d in reader:
+        fda_reg_dict[d['nct_id']] = d['is_fda_regulated']
+
+
+def is_covered_phase(phase):
+    return phase in ['Phase 1/Phase 2', 'Phase 2', 'Phase 2/Phase 3' , 'Phase 3', 'Phase 4', 'N/A']
+
+def is_not_withdrawn(study_status):
+    return study_status != 'Withdrawn'
+
+def is_interventional(study_type):
+    return study_type == 'Interventional'
+
+def is_covered_intervention(intervention_type):
+    for it in intervention_type:
+        if ('Drug' or 'Device' or 'Biological' or 'Genetic' or
```

Comment:
So, I sort of understand what you are saying, but I'm also a bit confused because I'm getting the right answer out the other end of the spreadsheet, where I would not be if this function wasn't operating as intended.

Either way, I understand how this approach works and as I am but a beginner I will absolutely be taking your word that this way is better. I've made this change to both this function and the `has_us_loc` function. The one difference is I added a `break` after finding a single `True` as in this instance, I don't care how many of the covered interventions or locations are listed as part of the trial, I just care about the first instance of any of them.

Reply:
Alright, so I was just testing this all out and this method breaks defining pACTs that rely on these two fields while my old method works BUT I think that's just based on some fluky luck with all the interventions I'm testing being Drugs and all the location countries being just United States. After some fiddling, I think I figured out the mechanics of all this and how to fix it so it will work in all instances:

What is going on here is that
`intervention_type = soup.find_all("intervention_type")`
is returning a list of bs4 tags that look like this:
`[<intervention_type>Drug</intervention_type>, <intervention_type>Drug</intervention_type>]`

Originally, we were looking for a string within those tags. That's a bit clunky so I'm adding this to turn just the text of each intervention type into a list.

```python
intervention_type_field = soup.find_all("intervention_type")
    trial_intervention_types = []
    for tag in intervention_type_field:
        trial_intervention_types.append(tag.get_text())
```

Then with some googling, I found a solution for comparing common elements in two lists with sets that I think is quite simple and elegant for this application, so I think I'm going to implement that:

```Python
def is_covered_intervention(intervention_type_list):
    covered_intervention_type = ['Drug','Device','Biological','Genetic','Radiation','Combination Prodcut','Diagnostic Test']
    a_set = set(covered_intervention_type)
    b_set = set(intervention_type_list)
    if (a_set & b_set):
        return True
    else:
        return False
```

Getting a US location is implemented differently as:

`locs = t(soup.location_countries)`

just gives you a big string of all the countries involved. Our test dataset only has trials with a single US location, so I went and downloaded some additional trial XMLs with multiple countries and played around with those to confirm this. It is indeed just one enormous string with all the country names separated by line breaks.

So, this is actually a bit easier to handle. We can just look for the strings of the US locations inside of that string:

```python
def has_us_loc(locs):
    us_locs = ['United States','American Samoa','Guam','Northern Mariana Islands','Puerto Rico','Virgin Islands (U.S.)']
    for us_loc in us_locs:
        if us_loc in locs:
            return True
            break
        else:
            return False
```

I've just made my included intervention types and country strings their own lists within the functions as I think this is more readable.

Reply:
When some "and" or "or" expression is evaluated, instead of making a new true or false value, Python uses the value that made the expression true or false. See it in the interactive interpreter (which is amazingly handy):


```
$ python3

Python 3.5.2 (default, Nov 12 2018, 13:43:14) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.

>>> ('Drug' or 'Device' or 'Biological' or 'Genetic' or 'Radiation' or 'Combination Prodcut' or 'Diagnostic Test')
'Drug'

>>> 'trueness' or ''
'trueness'

>>> 'trueness' and ''
''
```

If it was working, it must be that every study is a Drug study. I wanted you to see the operations and their orders. The first evaluation is list of ORs

`if ('Drug' or 'Device' or 'Biological' or 'Genetic' or 'Radiation' or 'Combination Prodcut' or 'Diagnostic Test') in it`

next becomes

`if ('Drug') in it`

which next becomes (because the parens around a simple expression is just that expression)

`if 'Drug' in it`

which next (depending on `it`) becomes

`if True`

which Python knows what to do with.